### PR TITLE
feat: add `BitVec.getMsbD_(add, sub, neg)` and `Bitvec.getLsbD(sub, neg)`

### PR DESCRIPTION
**Pull request by luisacicolini**
Replacing https://github.com/opencompl/lean-mlir/pull/686 which I mistakenly merged.

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
Why don't you want to merge this?

**Comment by luisacicolini** [comment]
I am not sure these are good enough, especially the `sub` ones

File: SSA/Projects/InstCombine/ForLean.lean [hunk 1]
```diff
@@ -574,6 +574,41 @@ theorem msb_sshiftRight' {x y: BitVec w} :
...
+theorem getMsbD_add {i : Nat} {i_lt : i < w} {x y : BitVec w} :
+    getMsbD (x + y) i =
+      Bool.xor (getMsbD x i) (Bool.xor (getMsbD y i) (carry (w - 1 - i) x y false)) := by
+  simp only [getMsbD]
+  by_cases h₀ : i < w
+  · simp only [h₀, decide_True, Bool.true_and]
+    rw [BitVec.getLsbD_add]
+    omega
+  · simp [h₀]
+    omega
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
  simp [getMsbD, getLsbD_add, i_lt, show w - 1 - i < w by omega]
```

File: SSA/Projects/InstCombine/ForLean.lean [hunk 2]
```diff
@@ -574,6 +574,41 @@ theorem msb_sshiftRight' {x y: BitVec w} :
     (x.sshiftRight' y).msb = x.msb := by
   simp [BitVec.sshiftRight', BitVec.sshiftRight_msb_eq_msb]
 
+theorem getLsbD_sub {i : Nat} {i_lt : i < w} {x y : BitVec w} :
+    (x - y).getLsbD i =
+      (x.getLsbD i ^^ ((~~~y + 1).getLsbD i ^^ carry i x (~~~y + 1) false)) := by
+  rw [BitVec.sub_eq_add_neg, BitVec.neg_eq_not_add]
+  rw [getLsbD_add]
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
  rw [BitVec.sub_eq_add_neg, BitVec.neg_eq_not_add, getLsbD_add]
```

File: SSA/Projects/InstCombine/ForLean.lean [hunk 3]
```diff
@@ -574,6 +574,41 @@ theorem msb_sshiftRight' {x y: BitVec w} :
...
+    omega
+  · simp [h₀]
+    omega
+
+theorem getMsbD_sub {i : Nat} {i_lt : i < w} {x y : BitVec w} :
+    (x - y).getMsbD i =
+      (x.getMsbD i ^^ ((~~~y + 1).getMsbD i ^^ carry (w - 1 - i) x (~~~y + 1) false)) := by
+  rw [BitVec.sub_eq_add_neg, BitVec.neg_eq_not_add]
+  rw [getMsbD_add]
+  omega
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
  simp [BitVec.sub_eq_add_neg, neg_eq_not_add, getMsbD_add]
```

**Review by tobiasgrosser — APPROVED** [review]
Otherwise, this looks great.

**Comment by tobiasgrosser** [comment]
@luisacicolini, this seems to be broken now. Can you double-check to enable the merge?

**Pull request merged**